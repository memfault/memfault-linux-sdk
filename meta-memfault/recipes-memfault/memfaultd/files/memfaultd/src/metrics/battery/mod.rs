//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
