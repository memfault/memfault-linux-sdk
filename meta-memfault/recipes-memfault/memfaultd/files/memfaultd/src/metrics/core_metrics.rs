//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";
