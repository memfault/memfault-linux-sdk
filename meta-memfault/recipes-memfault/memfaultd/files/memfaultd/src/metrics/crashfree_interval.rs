//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    cmp::max,
    iter::once,
    sync::mpsc::{channel, Receiver, Sender},
    time::Duration,
};

use chrono::Utc;
use tiny_http::{Method, Request, Response};

use crate::{
    http_server::{HttpHandler, HttpHandlerResult},
    metrics::{
        core_metrics::{
            METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
            METRIC_OPERATIONAL_HOURS,
        },
        MetricReading,
    },
    util::time_measure::TimeMeasure,
};

use super::{KeyedMetricReading, MetricStringKey};

pub struct CrashFreeIntervalTracker<T: TimeMeasure> {
    last_interval_mark: T,
    last_crashfree_interval_mark: T,
    crash_count: u32,
    sender: Sender<T>,
    receiver: Receiver<T>,
    interval: Duration,
    elapsed_intervals_key: MetricStringKey,
    crashfree_intervals_key: MetricStringKey,
    crash_count_key: MetricStringKey,
}

#[derive(Debug, PartialEq, Eq)]
struct TimeMod<T: TimeMeasure> {
    count: u32,
    mark: T,
}

impl<T> CrashFreeIntervalTracker<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync + 'static,
{
    pub fn new(
        interval: Duration,
        elapsed_intervals_key: MetricStringKey,
        crashfree_intervals_key: MetricStringKey,
        crash_count_key: MetricStringKey,
    ) -> Self {
        let (sender, receiver) = channel();
        Self {
            last_crashfree_interval_mark: T::now(),
            last_interval_mark: T::now(),
            sender,
            receiver,
            crash_count: 0,
            interval,
            elapsed_intervals_key,
            crashfree_intervals_key,
            crash_count_key,
        }
    }

    /// Returns a tracker with an hourly interval
    pub fn new_hourly() -> Self {
        Self::new(
            Duration::from_secs(3600),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        )
    }

    /// Wait for the next crash or update the metrics if the wait duration has passed.
    ///
    /// This allows us to have instant updates on crashes and hourly updates on the metrics, but
    /// also allows us to periodically update the metrics so that we don't have to wait for a crash.
    pub fn wait_and_update(&mut self, wait_duration: Duration) -> Vec<KeyedMetricReading> {
        if let Ok(crash_ts) = self.receiver.recv_timeout(wait_duration) {
            // Drain the receiver to get all crashes that happened since the last update
            self.receiver
                .try_iter()
                .chain(once(crash_ts))
                .for_each(|ts| {
                    self.crash_count += 1;
                    self.last_crashfree_interval_mark = max(self.last_crashfree_interval_mark, ts);
                });
        }

        // Since timing out just means no crashes occurred in the `wait_duration`,
        // update even when the receiver times out.
        self.update()
    }

    fn update(&mut self) -> Vec<KeyedMetricReading> {
        let TimeMod {
            count: count_op_interval,
            mark: last_counted_op_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_interval_mark);
        let TimeMod {
            count: count_crashfree_interval,
            mark: last_counted_crashfree_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_crashfree_interval_mark);

        self.last_interval_mark = last_counted_op_interval;
        self.last_crashfree_interval_mark = last_counted_crashfree_interval;

        let crashes = self.crash_count;
        self.crash_count = 0;

        let metrics_ts = Utc::now();
        vec![
            KeyedMetricReading::new(
                self.elapsed_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_op_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crashfree_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_crashfree_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crash_count_key.clone(),
                MetricReading::Counter {
                    value: crashes as f64,
                    timestamp: metrics_ts,
                },
            ),
        ]
    }

    pub fn http_handler(&mut self) -> Box<dyn HttpHandler> {
        Box::new(CrashFreeIntervalHttpHandler {
            channel: self.sender.clone(),
        })
    }

    pub fn capture_crash(&self) {
        self.sender
            .send(T::now())
            .expect("Failed to send crash timestamp");
    }

    /// Count how many `interval` have elapsed since `since`.
    ///
    /// This returns the number of intervals that have elapsed since `since`, and the timestamp of the end of the last interval
    /// that was counted. This is the value you should pass as `since` next time you call this function.
    ///
    /// See unit test for examples.
    fn full_interval_elapsed_since(interval: Duration, since: &T) -> TimeMod<T> {
        let now = T::now();
        if *since > now {
            return TimeMod {
                count: 0,
                mark: T::now(),
            };
        }

        let duration = now.since(since);
        let count_interval_elapsed = (duration.as_nanos() / interval.as_nanos()) as u32;
        TimeMod {
            count: count_interval_elapsed,
            mark: since.add(interval * count_interval_elapsed),
        }
    }
}

struct CrashFreeIntervalHttpHandler<T> {
    channel: Sender<T>,
}

impl<T> HttpHandler for CrashFreeIntervalHttpHandler<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync,
{
    fn handle_request(&self, request: &mut Request) -> HttpHandlerResult {
        if request.url() == "/v1/crash/report" && request.method() == &Method::Post {
            self.channel.send(T::now()).unwrap();
            HttpHandlerResult::Response(Response::from_string("OK").boxed())
        } else {
            HttpHandlerResult::NotHandled
        }
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use rstest::rstest;

    use crate::{
        metrics::{
            crashfree_interval::{
                METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
                METRIC_OPERATIONAL_HOURS,
            },
            KeyedMetricReading, MetricReading,
        },
        test_utils::TestInstant,
    };

    use super::CrashFreeIntervalTracker;
    use super::TimeMod;

    #[rstest]
    fn test_counting_intervals() {
        use std::time::Duration;

        // move the clock forward so we can go backwards below
        TestInstant::sleep(Duration::from_secs(3600));
        let now = TestInstant::now();

        let d10 = Duration::from_secs(10);
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(d10, &now),
            TimeMod {
                count: 0,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(10))
            ),
            TimeMod {
                count: 1,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(25))
            ),
            TimeMod {
                count: 2,
                mark: now - Duration::from_secs(5)
            }
        );
    }

    #[rstest]
    fn test_counting_hours() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(7200));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            2,
            2,
        );
    }

    #[rstest]
    fn test_counting_minutes() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new(
            Duration::from_secs(60),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        );

        TestInstant::sleep(Duration::from_secs(3600));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(3600));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            120,
            60,
        );
    }

    #[rstest]
    fn test_30min_heartbeat() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            1,
        );
    }

    #[rstest]
    fn test_30min_heartbeat_with_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        // Crash at t0 + 30min
        crashfree_tracker.capture_crash();

        // After 30' we should be ready to mark an operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );

        // After another 30' we should be ready to mark another crashfree hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            1,
        );

        // After another 30' we should be ready to mark another operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );
    }

    #[rstest]
    fn test_180min_heartbeat_with_one_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        // Basic test
        TestInstant::sleep(Duration::from_secs(3600 * 3));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            3,
        );

        // Crash at interval + 170'
        TestInstant::sleep(Duration::from_secs(170 * 60));
        crashfree_tracker.capture_crash();

        // Another 10' to the heartbeat mark
        // We will count 0 operational hour here. That is a consequence of the heartbeat being larger than the hour
        // To avoid this bug, we need to make sure we call the `update` at least once per hour!
        TestInstant::sleep(Duration::from_secs(10 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            0,
        );

        // However, doing the crash at interval +10' then waiting for 170' will record 2 crashfree hours
        TestInstant::sleep(Duration::from_secs(10 * 60));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(170 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            2,
        );
    }

    fn assert_operational_metrics(
        metrics: Vec<KeyedMetricReading>,
        expected_op_hours: u32,
        expected_crashfree_hours: u32,
    ) {
        assert_eq!(metrics.len(), 3);
        let op_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_HOURS)
            .unwrap();
        let crash_free_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_CRASHFREE_HOURS)
            .unwrap();

        let op_hours_value = match op_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        let crashfree_hours_value = match crash_free_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        assert_eq!(
            (op_hours_value as u32, crashfree_hours_value as u32),
            (expected_op_hours, expected_crashfree_hours)
        );
    }
}
