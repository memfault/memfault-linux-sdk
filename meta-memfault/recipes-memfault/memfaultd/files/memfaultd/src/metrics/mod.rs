//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
