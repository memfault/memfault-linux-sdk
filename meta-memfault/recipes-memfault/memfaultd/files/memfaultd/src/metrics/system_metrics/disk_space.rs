//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Collect disk space metric readings for devices listed in
//! /proc/mounts
//!
//! This module parses mounted devices and their mount points
//! from /proc/mounts and calculates how many bytes are free
//! and used on the device.
//!
use std::{
    collections::HashSet,
    fs::File,
    io::{BufRead, BufReader},
    path::{Path, PathBuf},
};

use eyre::{eyre, Result};
use log::warn;
use nix::sys::statvfs::statvfs;
use nom::{
    bytes::complete::take_while,
    character::complete::multispace1,
    sequence::{pair, preceded},
    IResult,
};

use serde::Serialize;

use crate::metrics::{system_metrics::SystemMetricFamilyCollector, KeyedMetricReading};

pub const DISKSPACE_METRIC_NAMESPACE_LEGACY: &str = "df";
pub const DISKSPACE_METRIC_NAMESPACE: &str = "disk_space";
pub const PROC_MOUNTS_PATH: &str = "/proc/mounts";

pub struct DiskSpaceInfo {
    block_size: u64,
    blocks: u64,
    blocks_free: u64,
}

#[cfg_attr(test, mockall::automock)]
pub trait DiskSpaceInfoForPath {
    fn disk_space_info_for_path(&self, p: &Path) -> Result<DiskSpaceInfo>;
}

pub struct NixStatvfs {}

impl NixStatvfs {
    pub fn new() -> Self {
        Self {}
    }
}

impl DiskSpaceInfoForPath for NixStatvfs {
    fn disk_space_info_for_path(&self, p: &Path) -> Result<DiskSpaceInfo> {
        let statfs = statvfs(p)
            .map_err(|e| eyre!("Failed to get statfs info for {}: {}", p.display(), e))?;

        Ok(DiskSpaceInfo {
            // Ignore unnecessary cast for these
            // as it is needed on 32-bit systems.
            #[allow(clippy::unnecessary_cast)]
            block_size: statfs.block_size() as u64,
            #[allow(clippy::unnecessary_cast)]
            blocks: statfs.blocks() as u64,
            #[allow(clippy::unnecessary_cast)]
            blocks_free: statfs.blocks_free() as u64,
        })
    }
}

#[derive(Serialize)]
struct Mount {
    device: PathBuf,
    mount_point: PathBuf,
}

pub enum DiskSpaceMetricsConfig {
    Auto,
    Disks(HashSet<String>),
}

pub struct DiskSpaceMetricCollector<T>
where
    T: DiskSpaceInfoForPath,
{
    config: DiskSpaceMetricsConfig,
    mounts: Vec<Mount>,
    disk_space_impl: T,
}

impl<T> DiskSpaceMetricCollector<T>
where
    T: DiskSpaceInfoForPath,
{
    pub fn new(disk_space_impl: T, config: DiskSpaceMetricsConfig) -> Self {
        Self {
            config,
            mounts: Vec::new(),
            disk_space_impl,
        }
    }
    fn disk_is_monitored(&self, disk: &str) -> bool {
        match &self.config {
            DiskSpaceMetricsConfig::Auto => disk.starts_with("/dev"),
            DiskSpaceMetricsConfig::Disks(configured_disks) => configured_disks.contains(disk),
        }
    }

    /// Parses a line of /proc/mounts for the name of
    /// the device the line corresponds to
    ///
    /// Example input:
    /// "/dev/sda2 / ext4 rw,noatime 0 0"
    /// Example output:
    /// "/dev/sda2"
    fn parse_proc_mounts_device(proc_mounts_line: &str) -> IResult<&str, &str> {
        take_while(|c: char| !c.is_whitespace())(proc_mounts_line)
    }

    /// Parses a line of /proc/mounts for the
    /// mount point the line corresponds to
    /// Parses /proc/mounts for a list of devices with active
    /// mount points in the system
    /// Example input:
    /// " / ext4 rw,noatime 0 0"
    /// Example output:
    /// "/"
    fn parse_proc_mounts_mount_point(proc_mounts_line: &str) -> IResult<&str, &str> {
        preceded(multispace1, take_while(|c: char| !c.is_whitespace()))(proc_mounts_line)
    }

    /// Parse a line of /proc/mounts
    /// Example input:
    /// "/dev/sda2 / ext4 rw,noatime 0 0"
    /// Example output:
    /// Mount { device: "/dev/sda2", "mount_point": "/" }
    fn parse_proc_mounts_line(line: &str) -> Result<Mount> {
        let (_remaining, (device, mount_point)) = pair(
            Self::parse_proc_mounts_device,
            Self::parse_proc_mounts_mount_point,
        )(line)
        .map_err(|e| eyre!("Failed to parse /proc/mounts line: {}", e))?;
        Ok(Mount {
            device: Path::new(device).to_path_buf(),
            mount_point: Path::new(mount_point).to_path_buf(),
        })
    }

    /// Initialize the list of mounted devices and their mount points based
    /// on the contents of /proc/mounts
    pub fn initialize_mounts(&mut self, proc_mounts_path: &Path) -> Result<()> {
        let file = File::open(proc_mounts_path)?;
        let reader = BufReader::new(file);

        for line in reader.lines() {
            // Discard errors - the assumption here is that we are only parsing
            // lines that follow the specified format and expect other lines in the file to error
            if let Ok(mount) = Self::parse_proc_mounts_line(line?.trim()) {
                if self.disk_is_monitored(&mount.device.to_string_lossy()) {
                    self.mounts.push(mount);
                }
            }
        }
        Ok(())
    }

    /// For a given mounted device, construct metric readings
    /// for how many bytes are used and free on the device
    fn get_metrics_for_mount(&self, mount: &Mount) -> Result<Vec<KeyedMetricReading>> {
        let mount_stats = self
            .disk_space_impl
            .disk_space_info_for_path(mount.mount_point.as_path())?;

        let block_size = mount_stats.block_size;
        let bytes_free = mount_stats.blocks_free * block_size;
        let bytes_used = (mount_stats.blocks * block_size) - bytes_free;

        let disk_id = mount
            .device
            .file_name()
            .ok_or_else(|| eyre!("Couldn't extract basename"))?
            .to_string_lossy();

        let bytes_free_reading = KeyedMetricReading::new_histogram(
            format!("disk_space/{}/free_bytes", disk_id)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key for bytes free: {}", e))?,
            bytes_free as f64,
        );

        let bytes_used_reading = KeyedMetricReading::new_histogram(
            format!("disk_space/{}/used_bytes", disk_id)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key for bytes used: {}", e))?,
            bytes_used as f64,
        );

        Ok(vec![bytes_free_reading, bytes_used_reading])
    }

    pub fn get_disk_space_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        if self.mounts.is_empty() {
            self.initialize_mounts(Path::new(PROC_MOUNTS_PATH))?;
        }

        let mut disk_space_readings = Vec::new();
        for mount in self.mounts.iter() {
            match self.get_metrics_for_mount(mount) {
                Ok(readings) => disk_space_readings.extend(readings),
                Err(e) => warn!(
                    "Failed to calculate disk space readings for {} mounted at {}: {}",
                    mount.device.display(),
                    mount.mount_point.display(),
                    e
                ),
            }
        }

        Ok(disk_space_readings)
    }
}

impl<T> SystemMetricFamilyCollector for DiskSpaceMetricCollector<T>
where
    T: DiskSpaceInfoForPath,
{
    fn family_name(&self) -> &'static str {
        DISKSPACE_METRIC_NAMESPACE
    }

    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        self.get_disk_space_metrics()
    }
}

#[cfg(test)]
mod test {
    use std::fs::File;
    use std::io::Write;

    use insta::{assert_json_snapshot, rounded_redaction};
    use rstest::rstest;
    use tempfile::tempdir;

    use super::*;

    #[rstest]
    fn test_process_valid_proc_mounts_line() {
        let line = "/dev/sda2 /media ext4 rw,noatime 0 0";
        let mount =
            DiskSpaceMetricCollector::<MockDiskSpaceInfoForPath>::parse_proc_mounts_line(line)
                .unwrap();

        assert_eq!(mount.device.as_os_str().to_string_lossy(), "/dev/sda2");
        assert_eq!(mount.mount_point.as_os_str().to_string_lossy(), "/media");
    }

    #[rstest]
    fn test_initialize_and_calc_disk_space_for_mounts() {
        let mut mock_statfs = MockDiskSpaceInfoForPath::new();

        mock_statfs
            .expect_disk_space_info_for_path()
            .times(2)
            .returning(|_p| {
                Ok(DiskSpaceInfo {
                    block_size: 4096,
                    blocks: 1024,
                    blocks_free: 286,
                })
            });

        let mut disk_space_collector =
            DiskSpaceMetricCollector::new(mock_statfs, DiskSpaceMetricsConfig::Auto);

        let line = "/dev/sda2 /media ext4 rw,noatime 0 0";
        let line2 = "/dev/sda1 / ext4 rw,noatime 0 0";

        let dir = tempdir().unwrap();

        let mounts_file_path = dir.path().join("mounts");
        let mut mounts_file = File::create(mounts_file_path.clone()).unwrap();

        writeln!(mounts_file, "{}", line).unwrap();
        writeln!(mounts_file, "{}", line2).unwrap();

        assert!(disk_space_collector
            .initialize_mounts(&mounts_file_path)
            .is_ok());

        disk_space_collector
            .mounts
            .sort_by(|a, b| a.device.cmp(&b.device));

        assert_json_snapshot!(disk_space_collector.mounts);

        let metrics = disk_space_collector.collect_metrics().unwrap();

        assert_json_snapshot!(metrics,
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)}
        );

        dir.close().unwrap();
    }
    #[rstest]
    fn test_unmonitored_disks_not_initialized() {
        let mock_statfs = MockDiskSpaceInfoForPath::new();

        let mut disk_space_collector = DiskSpaceMetricCollector::new(
            mock_statfs,
            DiskSpaceMetricsConfig::Disks(HashSet::from_iter(["/dev/sdc1".to_string()])),
        );

        let line = "/dev/sda2 /media ext4 rw,noatime 0 0";
        let line2 = "/dev/sda1 / ext4 rw,noatime 0 0";

        let dir = tempdir().unwrap();

        let mounts_file_path = dir.path().join("mounts");
        let mut mounts_file = File::create(mounts_file_path.clone()).unwrap();

        writeln!(mounts_file, "{}", line).unwrap();
        writeln!(mounts_file, "{}", line2).unwrap();

        assert!(disk_space_collector
            .initialize_mounts(&mounts_file_path)
            .is_ok());

        disk_space_collector
            .mounts
            .sort_by(|a, b| a.device.cmp(&b.device));

        assert!(disk_space_collector.mounts.is_empty());

        dir.close().unwrap();
    }
}
