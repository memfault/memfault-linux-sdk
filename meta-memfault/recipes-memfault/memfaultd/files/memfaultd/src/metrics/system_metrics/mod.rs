//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    thread::sleep,
    time::{Duration, Instant},
};

use eyre::Result;
use log::{debug, warn};

use crate::{
    config::SystemMetricConfig,
    metrics::KeyedMetricReading,
    util::system::{bytes_per_page, clock_ticks_per_second},
};

mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod thermal;
use crate::metrics::system_metrics::thermal::{ThermalMetricsCollector, THERMAL_METRIC_NAMESPACE};

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

mod network_interfaces;
use crate::metrics::system_metrics::network_interfaces::{
    NetworkInterfaceMetricCollector, NetworkInterfaceMetricsConfig,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
};

mod processes;
use processes::{ProcessMetricsCollector, PROCESSES_METRIC_NAMESPACE};

mod disk_space;
use disk_space::{
    DiskSpaceMetricCollector, DiskSpaceMetricsConfig, NixStatvfs, DISKSPACE_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
};

use self::processes::ProcessMetricsConfig;
use super::MetricsMBox;

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 7] = &[
    CPU_METRIC_NAMESPACE,
    MEMORY_METRIC_NAMESPACE,
    THERMAL_METRIC_NAMESPACE,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
    PROCESSES_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE,
    // Include in list of namespaces so that
    // legacy collectd from the "df" plugin
    // are still filtered out
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
];

pub trait SystemMetricFamilyCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>>;
    fn family_name(&self) -> &'static str;
}

pub struct SystemMetricsCollector {
    metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>>,
    metrics_mbox: MetricsMBox,
}

impl SystemMetricsCollector {
    pub fn new(system_metric_config: SystemMetricConfig, metrics_mbox: MetricsMBox) -> Self {
        // CPU, Memory, and Thermal metrics are captured by default
        let mut metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>> = vec![
            Box::new(CpuMetricCollector::new()),
            Box::new(MemoryMetricsCollector::new()),
            Box::new(ThermalMetricsCollector::new()),
        ];

        // Check if process metrics have been manually configured
        match system_metric_config.processes {
            Some(processes) if !processes.is_empty() => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Processes(processes),
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
            // Monitoring no processes means this collector is disabled
            Some(_empty_set) => {}
            None => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Auto,
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
        };

        // Check if disk space metrics have been manually configured
        match system_metric_config.disk_space {
            Some(disks) if !disks.is_empty() => {
                metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                    NixStatvfs::new(),
                    DiskSpaceMetricsConfig::Disks(disks),
                )))
            }
            // Monitoring no disks means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                NixStatvfs::new(),
                DiskSpaceMetricsConfig::Auto,
            ))),
        };

        // Check if network interface metrics have been manually configured
        match system_metric_config.network_interfaces {
            Some(interfaces) if !interfaces.is_empty() => metric_family_collectors.push(Box::new(
                NetworkInterfaceMetricCollector::<Instant>::new(
                    NetworkInterfaceMetricsConfig::Interfaces(interfaces),
                ),
            )),
            // Monitoring no interfaces means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(NetworkInterfaceMetricCollector::<
                Instant,
            >::new(
                NetworkInterfaceMetricsConfig::Auto
            ))),
        };

        Self {
            metric_family_collectors,
            metrics_mbox,
        }
    }

    pub fn run(&mut self, metric_poll_duration: Duration) {
        loop {
            for collector in self.metric_family_collectors.iter_mut() {
                match collector.collect_metrics() {
                    Ok(readings) => {
                        if let Err(e) = self.metrics_mbox.send_and_forget(readings) {
                            debug!(
                                "Couldn't add metric reading for family \"{}\": {:?}",
                                collector.family_name(),
                                e
                            )
                        }
                    }
                    Err(e) => warn!(
                        "Failed to collect readings for family \"{}\": {}",
                        collector.family_name(),
                        e
                    ),
                }
            }

            sleep(metric_poll_duration);
        }
    }
}
