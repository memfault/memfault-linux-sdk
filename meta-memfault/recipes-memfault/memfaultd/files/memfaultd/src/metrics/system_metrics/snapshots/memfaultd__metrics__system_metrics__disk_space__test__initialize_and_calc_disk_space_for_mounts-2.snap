---
source: memfaultd/src/metrics/system_metrics/disk_space.rs
expression: metrics
---
[
  {
    "name": "disk_space/sda1/free_bytes",
    "value": {
      "Histogram": {
        "value": 1171456.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "disk_space/sda1/used_bytes",
    "value": {
      "Histogram": {
        "value": 3022848.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "disk_space/sda2/free_bytes",
    "value": {
      "Histogram": {
        "value": 1171456.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "disk_space/sda2/used_bytes",
    "value": {
      "Histogram": {
        "value": 3022848.0,
        "timestamp": "[timestamp]"
      }
    }
  }
]
